### PR TITLE
Add Vec2, Rect and Mat3 header files

### PR DESCRIPTION
**Pull request by BenjaminPraud**

### CODE REVIEW

File: lib/include/SFVL/Mat3.hpp [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+#include <glm/glm.hpp>
+#include <glm/gtx/string_cast.hpp>
+
+#include <cmath>
+#include <iostream>
+
+namespace sfvl {
+template<typename T>
+
+struct Mat3
```
Comment (rmaxi-me) [inline lib/include/SFVL/Mat3.hpp]:
```suggestion
template<typename T>
struct Mat3
```
Extra line

File: lib/include/SFVL/Mat3.hpp [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+
+    constexpr Mat3() = default;
+    constexpr Mat3(T value) : m{std::array<T, 9>{value, 0.0, 0.0, 0.0, value, 0.0, 0.0, 0.0, value}} {}
+    constexpr Mat3(const std::array<T, 9> mat) : m{mat} {}
+    constexpr Mat3(T a1, T a2, T a3, T a4, T a5, T a6, T a7, T a8, T a9)
+        : m{std::array<T, 9>{a1, a2, a3, a4, a5, a6, a7, a8, a9}}
+    {
+    }
+
+    constexpr Mat3 inverse() noexcept
```
Comment (rmaxi-me) [inline lib/include/SFVL/Mat3.hpp]:
```suggestion
    constexpr Mat3 inverse() const noexcept
```
If your method does not change the class members, you should mark it as `const` so the if you have a `const Mat3` in your code, you can call the member function (among other things like boilerplate code and optimizations :))

File: lib/include/SFVL/Mat3.hpp [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+    {
+        auto cfMatrix = coFactor();
+        auto d = determinant();
+
+        return Mat3{cfMatrix.m[0] / d, cfMatrix.m[1] / d, cfMatrix.m[2] / d, cfMatrix.m[3] / d, cfMatrix.m[4] / d,
+                    cfMatrix.m[5] / d, cfMatrix.m[6] / d, cfMatrix.m[7] / d, cfMatrix.m[8] / d};
+    }
+
+    constexpr Mat3 coFactor() noexcept
```
Comment (rmaxi-me) [inline lib/include/SFVL/Mat3.hpp]:
```suggestion
    constexpr Mat3 coFactor() const noexcept
```

File: lib/include/SFVL/Mat3.hpp [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+    {
+        return Mat3{
+          ((m[4] * m[8]) - (m[5] * m[7])),  -((m[3] * m[8]) - (m[5] * m[6])), ((m[3] * m[7]) - (m[4] * m[6])),
+          -((m[1] * m[8]) - (m[2] * m[7])), ((m[0] * m[8]) - (m[2] * m[6])),  -((m[0] * m[7]) - (m[1] * m[6])),
+          ((m[1] * m[5]) - (m[2] * m[4])),  -((m[0] * m[5]) - (m[2] * m[3])), ((m[0] * m[4]) - (m[1] * m[3]))};
+    }
+
+    constexpr T determinant() noexcept
```
Comment (rmaxi-me) [inline lib/include/SFVL/Mat3.hpp]:
```suggestion
    constexpr T determinant() const noexcept
```

File: lib/include/SFVL/Mat3.hpp [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+    {
+        auto cfMatrix = coFactor();
+        return ((m[0] * cfMatrix.m[0]) + (m[1] * cfMatrix.m[1]) + (m[2] * cfMatrix.m[2]));
+    }
+
+    constexpr friend std::ostream& operator<<(std::ostream& os, const Mat3 mat3) noexcept
```
Comment (rmaxi-me) [inline lib/include/SFVL/Mat3.hpp]:
```suggestion
    constexpr friend std::ostream& operator<<(std::ostream& os, const Mat3& mat3) noexcept
```

File: lib/include/SFVL/Rect.hpp [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+    T left;
+    T top;
+    T width;
+    T height;
+
+    constexpr Rect() = default;
+    constexpr Rect(T aLeft, T aTop, T aWidth, T aHeight) : left{aLeft}, top{aTop}, width{aWidth}, height{aHeight} {}
+    constexpr Rect(const Vec2<T>& pos, const Vec2<T>& size) : left{pos.x}, top{pos.y}, width{size.x}, height{size.y} {}
+
+    constexpr bool operator==(const Rect<T>& right)
```
Comment (rmaxi-me) [inline lib/include/SFVL/Rect.hpp]:
```suggestion
    constexpr bool operator==(const Rect<T>& right) const noexcept
```

File: lib/include/SFVL/Rect.hpp [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+    {
+        return left == right.left && top == right.top && width == right.width && height == right.height;
+    }
+
+    constexpr bool operator!=(const Rect<T>& right)
```
Comment (rmaxi-me) [inline lib/include/SFVL/Rect.hpp]:
```suggestion
    constexpr bool operator!=(const Rect<T>& right) const noexcept
```

File: lib/include/SFVL/Rect.hpp [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+    {
+        return left != right.left || top != right.top || width != right.width || height != right.height;
+    }
+
+    [[nodiscard]] constexpr bool contains(T x, T y) const noexcept
+    {
+        if (left < x && top > y)
+            return true;
+        return false;
```
Comment (rmaxi-me) [inline lib/include/SFVL/Rect.hpp]:
```suggestion
        return left < x && top > y;
```
You can return it directly. Also, does this not take into account the width and height?

File: lib/include/SFVL/Rect.hpp [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+    }
+
+    [[nodiscard]] constexpr bool contains(const Vec2<T>& point) const noexcept
+    {
+        if (left < point.x && top > point.y)
+            return true;
+        return false;
```
Comment (rmaxi-me) [inline lib/include/SFVL/Rect.hpp]:
```suggestion
        return left < point.x && top > point.y;
```
Ditto

**Review by rmaxi-me — CHANGES_REQUESTED** [review]
Nice one :)

Just a few small adjustments

File: lib/include/SFVL/Rect.hpp (hunk 8 above)
Comment (BenjaminPraud) [reply lib/include/SFVL/Rect.hpp]: Yeah that's true I forgot these, I will fix that in the next commit

**Review by rmaxi-me — APPROVED** [review]
Resolve conflicts, wait for the CI, and 🚀

**Pull request merged**